napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import RLock
+from typing import Dict, Iterable, Optional, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+logger = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    """
+    High level class to control the creation of a slice (via a slice request),
+    submit it (synchronously or asynchronously) to a thread pool, and emit the
+    results when complete.
+
+    Events
+    ------
+    ready
+        emitted after slicing is done with a dict value that maps from layer
+        to slice response. Note that this may be emitted on a main or
+        non-main thread. If usage of this event relies on something happening
+        on the main thread, actions should be taken to ensure that the callback
+        is also executed on the main thread (e.g. by decorating the callback
+        with `ensure_main_thread`.
```

Comment:
```suggestion
        emitted after slicing is done with a dict value that maps from layer
        to slice response. Note that this may be emitted on the main or
        a non-main thread. If usage of this event relies on something happening
        on the main thread, actions should be taken to ensure that the callback
        is also executed on the main thread (e.g. by decorating the callback
        with `@ensure_main_thread`).
```

Otherwise, looks great to me.